src/train.py

```diff
@@ -163,39 +261,86 @@ def optimizer_step(model, optimizer, args):
+        loss = tts_loss + asr_loss + t_d_loss + s_d_loss
+    else:
+        asr_loss, tts_loss = supervised_step(model, batch)
+        loss = tts_loss + asr_loss
+    loss = loss / accum_steps
+
     loss.backward()
 
     # Log losses
-    losses['asr_'].append(asr_loss.detach().item())
-    losses['tts_'].append(tts_loss.detach().item())
+    losses['asr_'].append(asr_loss.detach().cpu().item())
+    losses['tts_'].append(tts_loss.detach().cpu().item())
+    if use_discriminator:
+        losses['t_sp_d'].append(t_d_loss.detach().cpu().item())
+        losses['s_sp_d'].append(s_d_loss.detach().cpu().item())
     return loss
 
-def train_ae_step(losses, model, batch, accum_steps):
+def train_ae_step(losses, model, batch, accum_steps, use_discriminator):
     batch = process_batch(batch)
-    t_ae_loss, s_ae_loss = autoencoder_step(model, batch)
-    loss = (t_ae_loss + s_ae_loss) / accum_steps
+    if use_discriminator:
+        t_ae_loss, s_ae_loss, t_d_loss, s_d_loss = autoencoder_step(model, batch, use_discriminator)
+        loss = t_ae_loss + s_ae_loss + t_d_loss + s_d_loss
+    else:
+        t_ae_loss, s_ae_loss = autoencoder_step(model, batch)
+        loss = t_ae_loss + s_ae_loss
+    loss = loss / accum_steps
     loss.backward()
 
     # Log losses
-    losses['t_ae'].append(t_ae_loss.detach().item())
-    losses['s_ae'].append(s_ae_loss.detach().item())
+    losses['t_ae'].append(t_ae_loss.detach().cpu().item())
+    losses['s_ae'].append(s_ae_loss.detach().cpu().item())
+    if use_discriminator:
+        losses['t_ae_d'].append(t_d_loss.detach().cpu().item())
+        losses['s_ae_d'].append(s_d_loss.detach().cpu().item())
     return loss
 
-def train_cm_step(losses, model, batch, accum_steps):
+def train_cm_step(losses, model, batch, accum_steps, use_discriminator):
+    # NOTE: do not use cross_model here bc need to take optimizer step inbetween
     batch = process_batch(batch)
-    t_cm_loss, s_cm_loss = crossmodel_step(model, batch)
-    loss = (s_cm_loss + t_cm_loss) / accum_steps
+
+    if use_discriminator:
+        t_cm_loss, s_cm_loss, s_d_loss, cm_t_d_loss, t_d_loss, cm_s_d_loss = crossmodel_step(model, batch, use_discriminator)
+        loss = s_cm_loss + t_cm_loss + s_d_loss + cm_t_d_loss + t_d_loss + cm_s_d_loss
+    else:
+        t_cm_loss, s_cm_loss = crossmodel_step(model, batch)
+        loss = s_cm_loss + t_cm_loss
+    loss = loss / accum_steps
     loss.backward()
 
     # Log losses
-    losses['s_cm'].append(s_cm_loss.detach().item())
-    losses['t_cm'].append(t_cm_loss.detach().item())
+    losses['s_cm'].append(s_cm_loss.detach().cpu().item())
+    losses['t_cm'].append(t_cm_loss.detach().cpu().item())
+    if args.use_discriminator:
+        losses['s_cm_d'].append(s_d_loss.detach().cpu().item())
+        losses['t_cm_d'].append(t_d_loss.detach().cpu().item())
+        losses['cm_s_cm_d'].append(cm_s_d_loss.detach().cpu().item())
```

Comment:
In keeping with there should be no gradients for the first step of the cross models (so text encode speech decode for text cross model and similarly for speech cross model) I think we shouldn't have this many discriminator losses...  Should we cut the loss from the first encoder?